src/App.tsx

```diff
@@ -1,26 +1,241 @@
+  deleteTodo,
+  getTodos,
+  updateTodo,
+} from './api/todos';
+import { ErrorNotification } from './components/errorNotification';
+import { Footer } from './components/footer';
+import { Header } from './components/header';
+import { TodoList } from './components/todoList';
+import { Todo } from './types/Todo';
 import { UserWarning } from './UserWarning';
 
-const USER_ID = 0;
-
 export const App: React.FC = () => {
-  if (!USER_ID) {
-    return <UserWarning />;
-  }
+  const [todos, setTodos] = useState<Todo[]>([]);
+  const [tempTodo, setTempTodo] = useState<Todo | null>(null);
+  const [newTodo, setNewTodo] = useState('');
+  const [error, setError] = useState<string | null>(null);
+  const [filter, setFilter] = useState<'all' | 'active' | 'completed'>('all');
+  const [loading, setLoading] = useState<boolean>(false);
+  const [editingTodoId, setEditingTodoId] = useState<number | null>(null);
+  const [editingTodoTitle, setEditingTodoTitle] = useState<string>('');
+
+  useEffect(() => {
+    if (USER_ID) {
+      setLoading(true);
+      getTodos()
+        .then(setTodos)
+        .catch(() => setError('Unable to load todos'))
+        .finally(() => setLoading(false));
+    }
+  }, []);
+
+  const handleAddTodo = (event: React.FormEvent) => {
+    event.preventDefault();
+    if (!newTodo.trim()) {
+      setError('Title should not be empty');
+
+      return;
+    }
+
+    setLoading(true);
+
+    const newTempTodo: Todo = {
+      id: 0,
+      title: newTodo,
+      completed: false,
+      userId: USER_ID,
+    };
+
+    setTempTodo(newTempTodo);
+
+    createTodo(newTodo)
+      .then((newTodoItem: Todo) => {
+        setTodos(prevTodos => [...prevTodos, newTodoItem]);
+        setNewTodo('');
+        setTempTodo(null);
+      })
+      .catch(() => {
+        setError('Unable to add a todo');
+        setTempTodo(null);
+      })
+      .finally(() => setLoading(false));
+  };
+
+  const handleToggleTodo = (todo: Todo) => {
+    const updatedTodo = { ...todo, completed: !todo.completed };
+
+    setLoading(true);
+
+    updateTodo(todo.id, updatedTodo)
+      .then(() => {
+        setTodos(prevTodos =>
+          prevTodos.map(t => (t.id === todo.id ? updatedTodo : t)),
+        );
+      })
+      .catch(() => {
+        setError('Unable to update a todo');
+      })
+      .finally(() => setLoading(false));
+  };
+
+  const handleToggleAllTodos = () => {
+    const allCompleted = todos.every(todo => todo.completed);
+    const newStatus = !allCompleted;
+
+    setLoading(true);
+
+    const updatedTodos = todos.map(todo => ({
+      ...todo,
+      completed: newStatus,
+    }));
+
+    Promise.all(
+      updatedTodos.map(todo =>
+        updateTodo(todo.id, todo).catch(() => {
+          setError('Unable to update a todo');
+        }),
+      ),
+    )
+      .then(() => {
+        setTodos(updatedTodos);
+      })
+      .finally(() => setLoading(false));
+  };
+
+  const handleDeleteTodo = (todoId: number) => {
```

Comment:
You should show the loader when adding/deleting (editing in the next task) todo
See working example and description in the task instructions

P.S. it also related to multiple todo actions such as `Clear completed` button

Reply:
Not fixed:
![image](https://github.com/user-attachments/assets/f35a4a1d-e13c-439d-8030-d413b05e7ffb)